### PR TITLE
feat(itemId): mirror list writes for itemId overflow prevention

### PR DESCRIPTION
**Pull request by kschelonka**
Mirror writes to list table. This is not the most efficient way to do it from a database perspective (adds an additional fetch to all updates), but is probably more efficient/safe from a code-change perspective. New flow:
* wrap all inserts/updates in a transaction context
* fetch the inserted/updated row
* write the updated row to the new "shadow table"
* close transaction

I probably didn't need to update the PocketSave code paths, but there were technically some small amount of usage I could see in apollo studio within 90 days (idk if test or what), so I just went ahead and did it.

**NOTE: This feature is behind a feature flag and will not be enabled upon merge.**

See also for Web repo proposed implementation https://github.com/Pocket/Web/pull/4615/

[Design doc](https://docs.google.com/document/d/1SqZVQ3D1GtzlfydbUFzx3ekbyACf3AOSuMEiEebtAJQ/edit#heading=h.gv486ynb1brr)

Prerequisites:
- [x] feature flag exists (and is disabled)
- [x] API keys for unleash access created and uploaded to AWS secrets manager
- [x] dev QA

[POCKET-9161]

[POCKET-9161]: https://mozilla-hub.atlassian.net/browse/POCKET-9161?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .aws/src/main.ts [hunk 1]
```diff
@@ -318,6 +322,10 @@ class ListAPI extends TerraformStack {
               name: 'DATABASE_WRITE_PASSWORD',
               valueFrom: databaseSecretEnvVars.writePassword,
             },
+            {
+              name: 'UNLEASH_KEY',
+              valueFrom: `arn:aws:secretsmanager:${region.name}:${caller.accountId}:secret:${config.name}/${config.environment}/UNLEASH_KEY`,
```
Comment (bassrock) [inline .aws/src/main.ts:327]: oo owe have keys!?

File: src/dataService/pocketSavesService.ts [hunk 2]
```diff
@@ -4,59 +4,14 @@ import { mysqlDateConvert, mysqlTimeString, setDifference } from './utils';
...
-  favorite: FavoriteStatus;
-  time_favorited: string; // Timestamp string
-};
-
-export enum FavoriteStatus {
-  FAVORITE = 1,
-  UNFAVORITE = 0,
-}
-
+import { SavedItemDataService } from './savedItemsService';
```
Comment (kschelonka) [inline src/dataService/pocketSavesService.ts:7]: Moved these types to a separate file for reuse

File: src/dataService/pocketSavesService.ts [hunk 3]
```diff
@@ -117,10 +72,25 @@ export class PocketSaveDataService {
...
+  constructor(
+    private context: Pick<
+      IContext,
+      'apiId' | 'dbClient' | 'userId' | 'unleash'
+    >,
+  ) {
     this.apiId = context.apiId;
     this.db = context.dbClient;
     this.userId = context.userId;
+    this.flags = {
```
Comment (kschelonka) [inline src/dataService/pocketSavesService.ts:87]: I found msyelf doing that multi-line isEnabled everywhere so I just simplified it here

File: src/dataService/savedItemsService.ts [hunk 4]
```diff
@@ -36,31 +37,53 @@ export class SavedItemDataService {
...
+        config.unleash.flags.mirrorWrites.name,
+        undefined,
+        config.unleash.flags.mirrorWrites.fallback,
+      ),
+    };
   }
 
-  public static convertDbResultStatus(dbResult: DbResult): DbResult;
-  public static convertDbResultStatus(dbResult: DbResult[]): DbResult[];
+  public static convertDbResultStatus<T extends Pick<RawListResult, 'status'>>(
```
Comment (kschelonka) [inline src/dataService/savedItemsService.ts:60]: Getting more specific about the actual type requirements for this function

File: .docker/mysql/schema/list_update_bigint_ids.sql [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    time_favorited datetime                   not null,
+    api_id         mediumint unsigned         not null,
+    status         tinyint unsigned           not null,
+    favorite       tinyint unsigned default 0 not null,
+    api_id_updated mediumint unsigned         not null,
+    primary key (user_id, item_id)
+)
+    collate = utf8_unicode_ci;
+
+--  Also create indices, so as to keep them updated while backfilling.
```
Comment (bassrock) [inline .docker/mysql/schema/list_update_bigint_ids.sql:24]: Is this cause not have the same index names?

File: src/config/index.ts [hunk 6]
```diff
@@ -125,6 +125,18 @@ export default {
...
   },
+  unleash: {
+    clientKey: process.env.UNLEASH_KEY || 'unleash-key-fake',
+    endpoint: process.env.UNLEASH_ENDPOINT || 'https://unleash-client.com/api',
+    flags: {
+      mirrorWrites: {
+        name: 'temp.backend.list_table_mirror_writes_enabled',
+        // TODO(@kschelon): Change this before rollout
+        // POCKET-9216
+        fallback: false,
```
Comment (bassrock) [inline src/config/index.ts:136]: easy config. love it.

**Review by bassrock — APPROVED** [review]
This is amazing. SO good.

File: .docker/mysql/schema/list_update_bigint_ids.sql (hunk 5 above)
Comment (kschelonka) [reply .docker/mysql/schema/list_update_bigint_ids.sql:24]: I assume so -- this I just copied from Marion's web repo PR

File: .aws/src/main.ts (hunk 1 above)
Comment (kschelonka) [reply .aws/src/main.ts:327]: Yep! I made them and put them in 1password (and AWS )

**Comment by kschelonka** [comment]
Was approved before the package-lock conflict fix.

**Pull request merged**